### PR TITLE
Add CocoaPods (Podfile.lock) Dependency Reporter

### PR DESCRIPTION
**Pull request by geraldnash-cb**

### CODE REVIEW

File: lib/salus/scanners/report_podfile_lock.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      dependencies = nil
+      begin
+        dependencies = JSON.parse(shell_return.stdout)
+      rescue JSON::ParserError
+        err_msg = "Could not parse JSON returned by bin/parse_podfile_lock's stdout!"
+        report_stderr(err_msg)
+        report_error(err_msg)
+        return
+      end
+
```
[the remark this answers is not in the document]
Comment (geraldnash-cb) [reply lib/salus/scanners/report_podfile_lock.rb:25]: LGTM! Thanks for the feedback

[the remark this answers is not in the document]
Comment (geraldnash-cb) [reply lib/salus/scanners/report_podfile_lock.rb:25]: I'm unable to force `dependencies` to be `nil` to directly test the above, but I've also added a test case (that tests `parse_podfile_lock`) for a lock file that simply has a literal `null` value.

File: bin/parse_podfile_lock [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env ruby
+
+require 'json'
+
+dependency_metadata_regex = /-\s(?<pod>.*?)\s\((=|(~>))?\s?(?<version>(\d*\.?)+)\)/
```
Comment (ghbren) [inline bin/parse_podfile_lock:5]: I'm worried about this regex.  Does cocoa provide dependencies in structured formats like json?

Comment (geraldnash-cb) [reply bin/parse_podfile_lock:5]: Not that I know of. There's a `pod list` command that's supposed to list pods but I'm not able to get it to work locally with `Podfile.lock`s. There was also an individually maintained tool called [`cocoapods-dependencies`](https://github.com/segiddins/cocoapods-dependencies), but it's gone three years without an update now, so I'm not sure we should use it. I'm open to thoughts, though!

Comment (ghbren) [reply bin/parse_podfile_lock:5]: Is the lock file itself always a yaml? It looks like so from the lockfiles you added to specs.  If so, you can just call yaml parse to get a structured objected of the values.

Comment (ghbren) [reply bin/parse_podfile_lock:5]: Actually, please take a look at the `to_yaml` method here
https://www.rubydoc.info/github/CocoaPods/Core/Pod/Lockfile#to_yaml-instance_method

Comment (geraldnash-cb) [reply bin/parse_podfile_lock]: Thanks for pointing that out, I didn't notice it's valid YAML. It's now using YAML to parse the dependency list, but because the version of each dependency is in the same string as the name it's still using a (smaller) regex.

Difference is before the regex grabbed every dependency and version of the file. Now the file is parsed as YAML and regex is used only to extract the name and version from each dependency element string.

File: bin/parse_podfile_lock [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env ruby
+
+require 'json'
+require 'yaml'
+
+dependency_metadata_regex = /(?<pod>.*?)\s\((=|(~>))?\s?(?<version>(\d*\.?)+)\)/
+
+podfile_lock = nil
```
Comment (saikat056) [inline bin/parse_podfile_lock:8]: We can write this as follows to avoid assigning `nil` to podfile_lock

```
podfile_lock = begin
  YAML.safe_load(File.read(ARGV[0]))
rescue SyntaxError, Psych::SyntaxError
  warn 'Unable to parse Podfile.lock file'
  exit(1)
end
```

Following is a quick test.
```
irb(main):001:1* a = begin
irb(main):002:1*   4
irb(main):003:1* rescue StandardError => e
irb(main):004:0> end
=> 4
irb(main):005:0> a
=> 4
irb(main):006:1* a = begin
irb(main):007:1*   raise StandardError.new
irb(main):008:1* rescue StandardError => e
irb(main):009:0> end
=> nil
irb(main):010:0> a
=> nil
```

File: bin/parse_podfile_lock [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  exit(1)
+end
+
+if podfile_lock['PODS'].nil? || podfile_lock['DEPENDENCIES'].nil?
+  warn '"PODS" or "DEPENDENCIES" key not found in Podfile.lock!'
+  exit(1)
+end
+
+dependencies = []
+# Parse individual dependency entry
```
Comment (saikat056) [inline bin/parse_podfile_lock:22]: My understanding is that the bodies of the both loops are same. Can we write it in the following way? Correct me if I am wrong.

```
(podfile_lock['PODS'] + podfile_lock['DEPENDENCIES']).each do |dependency_entry|
  dependency_properties = dependency_metadata_regex.match(dependency_entry)
  dependency_hash = {}
  dependency_hash[:pod] = dependency_properties[:pod]
  dependency_hash[:version] = dependency_properties[:version]
  dependencies.append(dependency_hash)
end
```

File: spec/fixtures/report_podfile_lock/bad_podfile_nil_value/Podfile.lock [hunk 5]
```diff
@@ -0,0 +1 @@
+null
```
Comment (saikat056) [inline spec/fixtures/report_podfile_lock/bad_podfile_nil_value/Podfile.lock:1]: **Nit**: Please add a new line in order to make GitHub happy.

File: spec/fixtures/report_podfile_lock/empty_podfile_values/Podfile.lock [hunk 6]
```diff
@@ -0,0 +1,7 @@
+PODS: 
+
+DEPENDENCIES: 
+
+PODFILE CHECKSUM: 
+
+COCOAPODS:
```
Comment (saikat056) [inline spec/fixtures/report_podfile_lock/empty_podfile_values/Podfile.lock:7]: **Nit**: Please add a new line in order to make GitHub happy.

File: spec/lib/salus/scanners/report_podfile_lock_spec.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+  it 'should report an error when an unparseable file is found' do
+    repo = Salus::Repo.new('spec/fixtures/report_podfile_lock/empty_podfile_values')
+    scanner = Salus::Scanners::ReportPodfileLock.new(repository: repo, config: {})
+
+    scanner.run
+
+    errs = scanner.report.to_h.fetch(:errors)
+    expect(errs.size).to eq(1)
+    expect(errs[0][:message]).to eq('"PODS" or "DEPENDENCIES" key not found in Podfile.lock!
```
Comment (saikat056) [inline spec/lib/salus/scanners/report_podfile_lock_spec.rb:89]: 👏 👏 Thank you for writing this test for bin/parse_podfile_lock.

**Comment by saikat056** [comment]
First of all, thank you for adding good test coverage to this pr. 🙏 

The change looks good to me. I added some minor suggestions. I want to see whether @ghbren or @joshua-ostrom has any other concerns. Otherwise, +1 from my side.

File: bin/parse_podfile_lock (hunk 3 above)
Comment (ghbren) [reply bin/parse_podfile_lock:8]: There is a `to_yaml` method here
https://www.rubydoc.info/github/CocoaPods/Core/Pod/Lockfile#to_yaml-instance_method

Gerald, could you look into whether you should use `to_yaml` or `YAML.safe_load`?

**Comment by ghbren** [comment]
> First of all, thank you for adding good test coverage to this pr. 🙏
> 
> The change looks good to me. I added some minor suggestions. I want to see whether @ghbren or @joshua-ostrom has any other concerns. Otherwise, +1 from my side.

Thanks for the review.  I haven't looked at all the code, but I think the regex definitely needs to be replaced with parsing yaml before we merge.

File: bin/parse_podfile_lock [hunk 8]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env ruby
+
+require 'json'
+require 'yaml'
+
+podfile_lock = begin
+  YAML.safe_load(File.read(ARGV[0]))
+               rescue SyntaxError, Psych::SyntaxError
```
Comment (ghbren) [inline bin/parse_podfile_lock:8]: `YAML.safe_load` throws more types of errors than the two above, like `Psych::DisallowedClass`, `Psych::BadAlias`, etc... 
It would be better to just `rescue StandardErrors`

File: lib/cyclonedx/report_podfile_lock_cyclonedx.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+module Cyclonedx
+  class ReportPodfileLock < Base
+    TYPE = "cocoa".freeze
```
Comment (ghbren) [inline lib/cyclonedx/report_podfile_lock_cyclonedx.rb:3]: It appears that this `TYPE` is not used?

File: lib/cyclonedx/report_podfile_lock_cyclonedx.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+    UNLOCKED_DEPENDENCY_FILE = "Podfile".freeze
```
Comment (ghbren) [inline lib/cyclonedx/report_podfile_lock_cyclonedx.rb:4]: ` UNLOCKED_DEPENDENCY_FILE` is also not used?

File: lib/cyclonedx/report_podfile_lock_cyclonedx.rb (hunk 9 above)
Comment (geraldnash-cb) [reply lib/cyclonedx/report_podfile_lock_cyclonedx.rb:3]: [It's used by the base class.](https://github.com/coinbase/salus/blob/8d4d7c8f80a0087b3e1fd1302be356d93369d842/lib/cyclonedx/base.rb#L42)

File: spec/lib/cyclonedx/report_podfile_lock_cyclonedx_spec.rb [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      repo = Salus::Repo.new('spec/fixtures/report_podfile_lock/normal')
+
+      scanner = Salus::Scanners::ReportPodfileLock.new(repository: repo, config: {})
+      scanner.run
+
+      cyclonedx_report = Cyclonedx::Report.new([[scanner.report, false]],
+                                               { "spec_version" => "1.3" })
+      cyclonedx_report_hash = cyclonedx_report.to_cyclonedx
+
+      expect { Cyclonedx::Report.validate_cyclonedx(cyclonedx_report_hash) }.not_to raise_error
```
[the remark this answers is not in the document]
Comment (ghbren) [reply spec/lib/cyclonedx/report_podfile_lock_cyclonedx_spec.rb:64]: Could you test what happens if with a lock file that cannot be parsed?
If you call `to_cyclonedx` on the result, will `validate_cyclonedx` still pass?

File: spec/lib/cyclonedx/report_podfile_lock_cyclonedx_spec.rb [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        [
+          {
+            "bom-ref": "pkg:cocoa/boost-for-react-native@1.63.0",
+            type: "library",
+            group: "",
+            name: "boost-for-react-native",
+            version: '1.63.0',
+            purl: "pkg:cocoa/boost-for-react-native@1.63.0",
+            properties: [
+              { key: "source", value: "" },
```
[the remark this answers is not in the document]
Comment (joshua-ostrom) [reply spec/lib/cyclonedx/report_podfile_lock_cyclonedx_spec.rb:23]: It would be nice to support the explicit source when given:

pod "Brightcove-Player-FreeWheel", :git => 'https://github.com/brightcove/BrightcoveSpecs.git'

Having that map to source would be a nice feature

File: spec/lib/cyclonedx/report_podfile_lock_cyclonedx_spec.rb [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+              { key: "dependency_file", value: "Podfile.lock" }
+            ]
+          },
+          {
+            "bom-ref": "pkg:cocoa/Flipper@%3D+0.87.0",
+            type: "library",
+            group: "",
+            name: "Flipper",
+            version: "= 0.87.0",
```
[the remark this answers is not in the document]
Comment (joshua-ostrom) [reply spec/lib/cyclonedx/report_podfile_lock_cyclonedx_spec.rb:44]: Yes let's have version limited to the version, qualifiers should be removed

File: bin/parse_podfile_lock [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  exit(1)
+end
+
+dependencies = []
+# Parse individual dependency entry
+(podfile_lock['PODS'] + podfile_lock['DEPENDENCIES']).each do |dependency_entry|
+  dependency_properties = dependency_entry.split(' (')
+  dependency_hash = {}
+  dependency_hash[:pod] = dependency_properties[0]
+  dependency_hash[:version] = dependency_properties[1].delete_suffix(')')
```
Comment (joshua-ostrom) [inline bin/parse_podfile_lock:24]: probably want to remove non numeric/alpha leading chars too.

File: bin/parse_podfile_lock [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+# Checks the PODS and DEPENDENCIES section
+podfile_dependencies = (lockfile.send :pod_versions)
+
+normalized_dependencies = []
+podfile_dependencies.each do |dependency_entry|
+  dependency_hash = {}
+  dependency_hash[:pod] = dependency_entry[0]
+  dependency_hash[:version] = dependency_entry[1]
+  normalized_dependencies.append(dependency_hash)
+end
```
Comment (ghbren) [inline bin/parse_podfile_lock:30]: Your code is fine here, but the can be updated to something like this to avoid accessing the array indexes.

```ruby
podfile_dependencies.each do |pod, version|
  dependency_hash = {}
  dependency_hash[:pod] = pod
  dependency_hash[:version] = version
  normalized_dependencies.append(dependency_hash)
end
```

**Review by ghbren — APPROVED** [review]
